Batch addition

moseschmiedel opened the pull request:
# Description
This implements batch addition of components for a given query. This is done by overloading the already existent `world.add` function with another version that takes a query instead of an entity (see `world.mojo` line 759ff).

# Default values for added components?
At the moment the function accepts default values for the added components. In #45 it is proposed to specify the added components only as parameters and leave component initialization to the user by iterating over all changed entities with the returned iterator. Should we do this?
As far as I can tell, the default setting of the component data impacts performance not significantly, as all entities need to be iterated anyway to add them to the new archetype and adjust the `EntityIndex` in the entity list.

moseschmiedel committed:
Add test case for batch component addition

moseschmiedel committed:
Draft batch component addition

moseschmiedel committed:
Archetype lookup and extension

moseschmiedel committed:
Fix batch add doc and test examples

moseschmiedel committed:
Overload `unsafe_set` to work with pointer to component data array

When doing batch operations it is more efficient to `memcpy` the whole component data array from the old archetype to the new archetype, instead of copying every component instance one by one.

moseschmiedel committed:
Overload `world.add` to accept a query for batch addition of components

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +756,144 @@ struct World[*component_types: ComponentType](
+                )
+
+                arch_start_idx = len(new_archetype[])
+                new_archetype[].reserve(arch_start_idx + len(old_archetype[]))
```
ensure that if a new allocation happens, the buffer size is increased by at least factor 2 so as to avoid frequent reallocations

samufi replied:
Maybe it makes sense to build this constraint into the reserve function... Not sure but could be a good idea.

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +756,144 @@ struct World[*component_types: ComponentType](
+                arch_start_idx = len(new_archetype[])
+                new_archetype[].reserve(arch_start_idx + len(old_archetype[]))
+
+                # save arch_start_idx for the iterator
```
I would start each inline coment with a capital letter unless it continues a sentence from the upper line

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +756,144 @@ struct World[*component_types: ComponentType](
+                new_archetype[].reserve(arch_start_idx + len(old_archetype[]))
+
+                # save arch_start_idx for the iterator
+                arch_start_idxs.append(arch_start_idx)
```
arch_start_idxs -> arch_start_idcs for better Engish spelling

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +756,144 @@ struct World[*component_types: ComponentType](
+
+                # add new components to the new archetype
+                @parameter
+                for comp_idx in range(add_components.__len__()):
```
I think you should be able to use `len()`

moseschmiedel committed:
Adjust mask and without_mask for returned iterator to match changed entities

moseschmiedel committed:
Use `world._get_entity_iterator`

moseschmiedel committed:
Test if archetypes have correct entity amount

moseschmiedel committed:
Debug printing

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +756,144 @@ struct World[*component_types: ComponentType](
+                    id = old_archetype[]._ids[i]
+
+                    new_archetype[].unsafe_set(
+                        arch_start_idx + i,
```
Why do you add the index of the component to the start index of the entity?

samufi replied:
In case this is an error, you could say `for id in old_archetype[]._component_count:`

moseschmiedel replied:
Yeah, this is mistake. This should only be `new_archetype[].unsafe_set(arch_start_idx, ...)`!

moseschmiedel commented:
Found a bug, where entities on the old archetype are not getting removed, therefore converted this to draft again.

moseschmiedel committed:
Align comment style with project style

moseschmiedel committed:
Don't add component idx to entity idx when copying component data

moseschmiedel committed:
Optimize `archetype.reserve` to reduce frequent reallocations

If we always allocate exactly the requested capacity, every call to `reserve` triggers a reallocation. To reduce the number of reallocations, we allocate twice the requested capacity. This way, the resulting `ArrayList` still has extra space available for use without requiring another full reallocation.

moseschmiedel committed:
Add StaticVariant

StaticVariant can model a value that can have multiple types but it is known at compile-time which type will be used.
It may be helpful to think of it as the Variant type from stdlib, but resolved completely at compile-time.

moseschmiedel committed:
Implement common traits for StaticVariant when wrapped type implements them

moseschmiedel committed:
StaticVariant copy constructor and other trait methods crash

See https://github.com/modular/modular/issues/5172

moseschmiedel committed:
Introduce ArchetypeListIterator to iterate over a given list of archetypes

In some operations we already know which archetypes where changed. Therefore it is more efficient to store a list of these archetypes and return an iterator that uses this list to iterate over all archetypes.
This can also be used for the _EntityIterator to also iterate through all entities of the iterated archetypes.

moseschmiedel committed:
ArchetypeListIterator should only hold one pointer to all archetypes in world

We cannot use a list of pointers in the ArchetypeListIterator, because then multiple mutable references to the same list could exist. This is not allowed in Mojo. Instead we store a list of indices into the global _archetype list of the world.

moseschmiedel committed:
Use ArchetypeListIterator for batch addition and clear old_archetype correctly

moseschmiedel committed:
Remove debug print

samufi reviewed (changes requested):
Great work, only small issues. Could you please also add a benchmark for batch add?

moseschmiedel committed:
Clarify docstring for `Archetype.reserve`

moseschmiedel committed:
Rename `ArchetypeMaskIterator` to `ArchetypeByMaskIterator`

moseschmiedel committed:
Use existing functionality for `Archetype.unsafe_set` for multiple entities

moseschmiedel committed:
Fix naming of ArchetypeIterators

moseschmiedel committed:
Remove unneeded code from `_ArchetypeIterators`

moseschmiedel committed:
Change `StaticVariant.isa` to `is_a` and `EltType` to `ElementType`

moseschmiedel committed:
Add comment about future fix with `requires` for `StaticVariant`

moseschmiedel committed:
Clarify comments for batch  `world.add`

moseschmiedel committed:
Rename `_ArchetypeByListIterator._buffer_index` to `_index`

moseschmiedel committed:
Preallocate space for `changed_archetype_idcs` and `arch_start_idcs`

moseschmiedel committed:
Fix naming of index of old archetype in `world.add`

moseschmiedel committed:
Fix pointer invalidation bug in `world.add`

moseschmiedel committed:
Fix batch `world.add` error message

moseschmiedel committed:
Change `test_world_batch_add` to operate on multiple entities

moseschmiedel committed:
Fix `test_world_batch_add` to expect correct error message

samufi reviewed (changes requested):
Could you add a benchmark of batch add?

moseschmiedel commented:
> Could you add a benchmark of batch add?

Sure.

moseschmiedel committed:
Add benchmarks for batch adding components to entities

moseschmiedel committed:
Fix duplication of entities during batch `world.add`

In the batched version of `world.add` an entity was added once per new component to the `new_archetype`. So if two new components were added the `new_archetype` was extended the amount of entities in `old_archetype` TIMES TWO. This should not happen as it will fill the available memory very quickly!

moseschmiedel committed:
Use entity methods in `test_world_batch_add` for component access

mlange-42 commented on src/larecs/archetype.mojo:
```diff
@@ -432,20 +432,23 @@ struct Archetype[
             id = self._ids[i]
             old_size = index(self._item_sizes[id]) * self._capacity
-            new_size = index(self._item_sizes[id]) * new_capacity
+            new_size = index(self._item_sizes[id]) * new_capacity * 2
```
Does it really make sense to increase the capacity here again? Isn't the calculation already done on the caller's side?

BTW, looking at the resp. calculation in line 732:

```mojo
new_capacity = max(self._size + count, UInt(2) * self._capacity)
```

we get "odd" values if the intended capacity is more than twice the current one. In Ark, I always grow to the next power-of-two value.

moseschmiedel replied:
I think I don't completely understand what you're suggesting. Do you think we should do the calculation inside `reserve` (like it is implemented in this PR) or increase the capacity only, so that the `new_capacity` fits?

mlange-42 replied:
In this PR, you set the capacity to twice the requested capacity (i.e. twice the one given as the argument), which does not really make sense IMO. The variant in line 732 looks better in that it doubles the current capacity if necessary. However, it does so only once, not until the new capacity matches or exceeds the requested one. So if the current capacity is 128 and you request 277, you get 277 instead of 512. And it does so outside of `reserve`, so the result gets doubled again (at least I think so, without looking at the code again).

My suggestion would be to do such calculations only in `reserve`, never outside of it, and to double the current (not the requested) capacity until sufficient, like this (or with some log2+pow calculation or bit-shift trick):

```mojo
if new_capacity <= self._capacity:
    return

cap = self._capacity
while cap < new_capacity:
    cap *= 2

# do the actual work...
```

mlange-42 commented on src/larecs/query.mojo:
```diff
@@ -392,22 +395,104 @@ struct _ArchetypeIterator[
+        Returns the number of archetypes remaining in the iterator.
+
+        Note that this requires iterating over all archetypes
+        and may be a complex operation.
```
Leftover from C&P iterator over entities?

mlange-42 commented on src/larecs/world.mojo:
```diff
@@ -756,6 +812,177 @@ struct World[*component_types: ComponentType](
+            )
+
+        arch_start_idcs = List[UInt, True](len(self._archetypes))
+        changed_archetype_idcs = List[Int, True](len(self._archetypes))
```
It might be overly pessimistic to reserve memory for the total number of archetypes in the world. Maybe set a (generous) upper limit and risk growing the list later?

moseschmiedel replied:
Yes, I was thinking that the total number of archetypes will never be that big. Even `255 * 8 byte ~ 2kb` which is nothing, keeping in mind that current L1 caches have hundreds of kilobyte (e.g. see https://www.amd.com/de/support/downloads/drivers.html/processors/ryzen/ryzen-5000-series/amd-ryzen-5-5600.html).
What would your proposed initial capacity be?

moseschmiedel committed:
Getting length of `_ArchetypeByListIterator` is not hard

moseschmiedel committed:
Add function to calculate next power of 2 fast

moseschmiedel committed:
If requested Archetype capacity exceeds available increase to next sufficient power of 2

moseschmiedel committed:
Change `Archetype.reserve` calls so that capacity calculation is handled by `reserve` and not outside of it

moseschmiedel committed:
Fix `next_pow2` to work generically for all `UInt`s

moseschmiedel committed:
Change `archetype_test` according to new reservation rules

moseschmiedel committed:
Only pre-allocate 2kb at maximum in batch `world.add`

moseschmiedel committed:
Add helper `QueryInfo.matches` to encapsulate query matching logic

moseschmiedel committed:
Use `QueryInfo.matches` in `_ArchetypeByMaskIterator`

moseschmiedel committed:
Refactor test for conflicting archetypes in batch `world.add` to check only existing archetypes

moseschmiedel committed:
Fix `QueryInfo.matches` and use it also for
`_ArchetypeByMaskIterator.__len__`

moseschmiedel committed:
Add `BitMask.__or__`

moseschmiedel committed:
Fix query check in batch `world.add` to search for archetypes that would match after component addition

moseschmiedel committed:
Extend `test_world_batch_add` to check that batch `world.add` doesn't throw an error when only creating archetypes not already existing

moseschmiedel committed:
Fix `world_benchmark.mojo`

moseschmiedel commented on src/larecs/world.mojo:
```diff
@@ -756,6 +812,177 @@ struct World[*component_types: ComponentType](
+        alias component_ids = Self.component_manager.get_id_arr[*Ts]()
+
+        # If query could match archetypes that already have at least one of the components, raise an error
+        if not query.without_mask[].contains(BitMask(component_ids)):
```
I implemented a naive check where all matching existing archetypes are iterated and checked if they have conflicting components. This is quite a bit more costly then the BitMask check (`O(n)` instead of `O(1)`) but it does not complain as often as the other implementation ...and it is still faster then individual `world.add` calls according to the benchmark.

moseschmiedel replied:
What are your thoughts on this implementation?

samufi replied:
@mlange-42 I do not understand how the stricter version could lead to faulty reasoning, as it requires a more explicit statement. Could you elaborate / give an example?
I think having the more expensive check is fine, but I would wrap it with a check of the simpler condition. Then people who want to be explicit can do that and save time that way.

moseschmiedel committed:
Update doc-strings

moseschmiedel committed:
Add doc section about batch component addition

mlange-42 commented on src/larecs/world.mojo:
```diff
@@ -756,6 +867,192 @@ struct World[*component_types: ComponentType](
+        for archetype in self._get_archetype_iterator(
+            query.mask, query.without_mask
+        ):
+            if query.matches(archetype[].get_mask()):
+                if not archetype[].get_mask().contains(
+                    BitMask(component_ids)
+                ) and archetype[].get_mask().contains_any(
+                    BitMask(component_ids)
+                ):
```
1. Isn't the outer condition (`query.matches(...)`) already covered by the iterator?
1. If the second part of the inner condition gets evaluated, you construct the mask twice.
1. I do not really understand the first part og the inner condition. Why is it ok that the archetype already contains all the components that are to be added?

A general comment/question: do you (particularly @samufi) assume that all raised errors can be caught and recovered? My approach in Ark is to panic on "misuse", which makes it easier to do checks like this "on the fly" and not needing to check everything before making changes.

mlange-42 replied:
Ok, this makes sense somehow.

But maybe it would be useful to add a user guide section on the "error philosophy"? Particulary stating that catching errors should not be a "usual thing", but that Larecs raises errors only on problems that are to be considered either misuse or mistakes in the model logoc.

moseschmiedel replied:
> 1. Isn't the outer condition (query.matches(...)) already covered by the iterator?
Yes, this a left-over from a previous idea I had...😅

> 2. If the second part of the inner condition gets evaluated, you construct the mask twice.
I think it wouldn't because the compiler could optimize that. See my comment on the message from @samufi 

> 3. I do not really understand the first part of the inner condition. Why is it ok that the archetype already contains all the components that are to be added?
It is ok, because then it is one of the target archetypes. It wouldn't lead to multiple source archetypes mapping on to one target archetype. Consider following example
```
Archetypes: A, AB, AC, ABC
add(query[A], B, C)
```
In this example archetype `ABC` is no problem because it is the target of `A + BC`. But `AB` and  `AC` are problems, because they would also get transformed to `ABC`, which leads to following mapping:
```
A -> ABC
AB -> ABC
AC -> ABC
```
We don't want this, because then we have dependencies between the loop iterations, which would prevent parallization later on.

mlange-42 replied:
So adding B+C to ABC is a no-op (or rather, replaces the components)?

IMO it is not a good idea to allow this, as it is counter-intuitive and potentially error-prone. Also, allowing to replace all but not allowing to replace some and add others seems inconsistent to me.

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +867,192 @@ struct World[*component_types: ComponentType](
+        ):
+            if query.matches(archetype[].get_mask()):
+                if not archetype[].get_mask().contains(
+                    BitMask(component_ids)
```
Is there really a need for a cast in each for loop iteration?

moseschmiedel replied:
No, but AFAIK this is really easy for the compiler to optimize and therefore there should be no need for manual optimization here.

samufi replied:
Interesting. How reliable are such things?

moseschmiedel committed:
Make cheap Bitmask check in `world.add` before iterating through provided query

moseschmiedel commented on src/larecs/world.mojo:
```diff
@@ -756,6 +867,192 @@ struct World[*component_types: ComponentType](
+                    BitMask(component_ids)
+                ):
+                    raise Error(
+                        "Query could match entities that already have at least"
```
No, I changed that!

moseschmiedel committed:
Fix `world_test` due to change of wording in error message

moseschmiedel commented on src/larecs/world.mojo:
```diff
@@ -756,6 +867,192 @@ struct World[*component_types: ComponentType](
+                        " components."
+                    )
+
+        alias _2kb = (1024 * 2) // sizeof[UInt]()
```
Is the name of the alias not descriptive enough for you? What sort of information do you want to have here as a comment?

moseschmiedel committed:
Add reminder to fix `has_without_mask` short-circuiting later

moseschmiedel committed:
Rewrite `_2kb` alias to be (hopefully) more descriptive

samufi commented on src/larecs/world.mojo:
```diff
@@ -756,6 +867,189 @@ struct World[*component_types: ComponentType](
+        alias component_ids = Self.component_manager.get_id_arr[*Ts]()
+
+        # If query could match archetypes that already have at least one of the components, raise an error
+        if not query.without_mask.or_else(BitMask()).contains(
```
Note, we are not working with Optional here but StaticOptional. Use an `@parameter if`, then this should work.

moseschmiedel replied:
No, then the second condition cannot be part of the if statement anymore, because it depends on a dynamic value.

samufi replied:
Couldn't you solve this via an intermediate boolean value, which is constant True if not has_without_mask and else-wise query.without_mask[].contains(...)? The extra value may be optimized away on the constant branch...

moseschmiedel replied:
Do you mean like that?
```suggestion
        foo = False
        
        @parameter
        if has_without_mask:
            foo = True

        if foo or not query.without_mask[].contains(
```

This triggers the same bug mentioned earlier.

moseschmiedel committed:
Improve 2kb name

moseschmiedel committed:
Refactor check if stricter mask check is needed

moseschmiedel committed:
Skip empty archetypes in strict check

moseschmiedel committed:
Change alias name to snake_case

moseschmiedel committed:
Merge if expressions
